### ISSUE TITLE
Reduce goroutine creation when cloning repos

### ISSUE THREAD

**Issue by indradhanush** [sourcegraph/sourcegraph#25031]
Currently when we receive requests to clone a new repo we spawn a goroutine and wait until we are allowed to clone based on the number of other concurrent clones happening. 

https://github.com/sourcegraph/sourcegraph/blob/1c9e6f78cc6494318a5e3974975aa996b80e69dc/cmd/gitserver/server/server.go#L1280

During rebalance operations this leads to a huge spike in go routines which consumes a lot of memory and introduces the danger that the gitserver pod is OOM killed.

Instead, we should create n goroutines on startup, where n is the number of concurrent clones allowed and add clone requests to an internal queue

**Comment by indradhanush** [comment]
We've merged the changes and will monitor the metrics once it is deployed to dotcom. If the metrics are able to verify that our approach is correct - we will create a new PR to enable this which would then allow this issue to be closed once the PR is merged.

**Comment by indradhanush** [comment]
We had to change the metrics from gauge to counters in https://github.com/sourcegraph/sourcegraph/pull/25287. And here are the screenshots of the metrics from a few randomly chosen gitserver instances. All the metrics **rise as one** and it verifies that our approach with the producer-consumer pipeline should not introduce any regressions. Next step is to enable the pipeline and remove the older approach.

#### gitserver-0
![image](https://user-images.githubusercontent.com/2682729/134865118-33fded7f-4a09-4839-8337-c3798e61e803.png)

#### gitserver-1
![image](https://user-images.githubusercontent.com/2682729/134865224-9a62a3b4-71dd-4b52-af37-a077e88e4a3f.png)

#### gitserver-7
![image](https://user-images.githubusercontent.com/2682729/134865446-5e281680-083e-423f-ba35-3299cb506959.png)

#### gitserver-13
![image](https://user-images.githubusercontent.com/2682729/134865543-cf23b75c-180e-4134-a0e2-04b4927b1edb.png)

#### gitserver-18
![image](https://user-images.githubusercontent.com/2682729/134865625-ca4d9b4a-044a-46b1-946c-24a59fcf51db.png)

#### gitserver-19
![image](https://user-images.githubusercontent.com/2682729/134865745-b6b1e2ec-d98c-40fa-ac75-8560a24520cf.png)
